### PR TITLE
Add base_asset to Market structure and make prediction_markets pallet use it instead of Asset::Ztg

### LINKED ISSUE #226: [PM]`PredictionMarkets` can use any approved token as a base currency
Currently, the `PredictionMarkets` pallet expects that only the native currency (ZTG) is being used as the base currency for all prediction markets to settle in. However, stablecoin like kUSD will also be a popular choice when creating markets, and should be able to be used as a base currency as well.

- There should be a list of approved tokens that can be used as base currency for prediction markets. The list of approved tokens can be changed by governance or the root origin. 
- We should add `USD` as a token type in the [`Asset`](https://github.com/zeitgeistpm/zeitgeist/blob/main/primitives/src/asset.rs#L22) enum, which currently is just stubbed.
- `PredictionMarkets` internal logic should be changed from assuming that all markets are using `ZTG` as the base currency, to using any of the approved assets.

### PR DESCRIPTION
**Pull request by vivekvpandya**
Fixes #226

### CODE REVIEW

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (vivekvpandya) [inline zrml/prediction-markets/src/orml_asset_registry.rs:1]: This code is almost taken as it is from Centrifuge repo. AFAIU we need to keep their license note right?

File: zrml/prediction-markets/src/tests.rs
Comment (Chralt98) [inline zrml/prediction-markets/src/tests.rs:379]: So `allow_in_pool` is meant to be a guarantee, that we allow the `true` assets as base asset in prediction markets? Shouldn't we rather add another boolean flag like `allow_as_base_asset`?

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (Chralt98) [reply zrml/prediction-markets/src/orml_asset_registry.rs:1]: Can you explain a little bit, what this file does and what it is needed for? Thanks!

File: zrml/market-commons/src/migrations.rs
Comment (Chralt98) [inline zrml/market-commons/src/migrations.rs:38]: Currently there are these derives for a market https://github.com/zeitgeistpm/zeitgeist/blob/f8f5c79c7c49c9011ab00ffa670d40dab48d44b7/primitives/src/market.rs#L30

**Review by Chralt98 — COMMENTED** [review]
Essentially this is a great work to allow different base assets in prediction markets. I think about unintended consequences.

File: zrml/prediction-markets/src/mock.rs
Comment (maltekliemann) [inline zrml/prediction-markets/src/mock.rs:420]: This should also be tagged with `#[cfg(feature = "parachainn")]`. Might be stylistically better to use https://docs.rs/cfg-if/0.1.2/cfg_if/ here.

File: zrml/prediction-markets/src/mock.rs
Comment (maltekliemann) [inline zrml/prediction-markets/src/mock.rs:30]: Should only be imported by `parachain` is activated.

**Review by maltekliemann — COMMENTED** [review]
Just passing through and leaving a couple of hopefully useful comments. Do you understand what the purpose of using the `__private` prefix on the imports in the `orml_mock_registry` module is?

File: zrml/prediction-markets/src/lib.rs
Comment (vivekvpandya) [reply zrml/prediction-markets/src/lib.rs:624]: I think we should add an error for UnregisteredAsset.

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (vivekvpandya) [reply zrml/prediction-markets/src/orml_asset_registry.rs:1]: No logical changes, just applied cargo fmt.

Comment (vivekvpandya) [reply zrml/prediction-markets/src/orml_asset_registry.rs:1]: @Chralt98 this is required to implement traits on a mock structure to be used for AssetRegistry config item.
@maltekliemann should we move it in future when its required outside pm pallet?

File: zrml/market-commons/src/migrations.rs
Comment (vivekvpandya) [reply zrml/market-commons/src/migrations.rs:38]: but do we really need other traits derived ? I have taken reference from https://github.com/zeitgeistpm/zeitgeist/pull/732/files#diff-7eef70e0e24e29abbdd8ae4c5949b3d7633f1fa7b844e9b0e7ff3227cc344865R50 and it works fine in the test migration.

File: primitives/src/types.rs
Comment (Chralt98) [inline primitives/src/types.rs:145]: Now the description seems a little bit off.

File: zrml/market-commons/src/migrations.rs
Comment (Chralt98) [reply zrml/market-commons/src/migrations.rs:38]: I must admit that I don't know an answer to this question. So if you have the proof that it works, we should ignore this.

File: zrml/prediction-markets/src/lib.rs
Comment (Chralt98) [inline zrml/prediction-markets/src/lib.rs:624]: That looks kind of strange. If `valid_base_asset` contains the Error `UnregisteredForeignAsset`, what does happen when this reaches the below ensure check? I think this is quite confusing. 

```suggestion
            let mut valid_base_asset = false;
            match base_asset {
                Asset::Ztg => valid_base_asset = true,
                #[cfg(feature = "parachain")]
                Asset::ForeignAsset(fa) => {
                    let metadata = T::AssetRegistry::metadata(&Asset::ForeignAsset(fa)).ok_or(Error::<T>::UnregisteredForeignAsset);
                    valid_base_asset = metadata..additional.allow_as_base_asset;
                }
                _ => valid_base_asset = false,
            }?;
```

**Review by Chralt98 — COMMENTED** [review]
Looks okay, when you ask me.

File: zrml/prediction-markets/src/lib.rs
Comment (Chralt98) [reply zrml/prediction-markets/src/lib.rs:624]: I am sorry. I did not see [that.](https://github.com/zeitgeistpm/zeitgeist/pull/894#discussion_r1048946784) But the style with error seems still a little bit strange.

Comment (vivekvpandya) [reply zrml/prediction-markets/src/lib.rs:624]: ? on line 622 will make function return error and it never reaches ensure on line 626. And test is updated to cover this case.

Comment (vivekvpandya) [reply zrml/prediction-markets/src/lib.rs:624]: earlier it was like what @Chralt98 suggested but after @maltekliemann 's suggestion I changed it. I find both options fine.

Comment (Chralt98) [reply zrml/prediction-markets/src/lib.rs:624]: It's just that I don't think that the following syntax is well understood at the first glance.

```rust
Ok::<bool, Error<T>>(true)
```

So I would like to stick with the "readable" version with `mut valid_base_asset`, but it's not that important. It just was so noticeable.

Comment (maltekliemann) [reply zrml/prediction-markets/src/lib.rs:624]: Chris is absolutely right, that's not pretty. But if you just add a type hint to `valid_base_asset`, we can get rid of the other type hints.

```rust
            let valid_base_asset: bool = match base_asset {
                Asset::Ztg => Ok(true),
                #[cfg(feature = "parachain")]
                Asset::ForeignAsset(fa) => {
                    if let Some(metadata) = T::AssetRegistry::metadata(&Asset::ForeignAsset(fa)) {
                        Ok(metadata.additional.allow_as_base_asset)
                    } else {
                        Err(Error::<T>::UnregisteredForeignAsset)
                    }
                }
                _ => Ok(false),
            }?;
```

Comment (maltekliemann) [reply zrml/prediction-markets/src/lib.rs:624]: You can also skip the entire result and just return an error where appropriate:

```rust
            let valid_base_asset = match base_asset {
                Asset::Ztg => true,
                #[cfg(feature = "parachain")]
                Asset::ForeignAsset(fa) => {
                    if let Some(metadata) = T::AssetRegistry::metadata(&Asset::ForeignAsset(fa)) {
                        metadata.additional.allow_as_base_asset
                    } else {
                        return Err(Error::<T>::UnregisteredForeignAsset.into());
                    }
                }
                _ => false,
            };
```

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (maltekliemann) [reply zrml/prediction-markets/src/orml_asset_registry.rs:18]: I'd put it in `primitives/` now, but it's fine to keep it here, as well.

**Review by maltekliemann — CHANGES_REQUESTED** [review]
Migration looks good. We need to fix how market bonds are unreserved and move a check to ensure that markets can't be edited in illegal ways.

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (vivekvpandya) [reply zrml/prediction-markets/src/orml_asset_registry.rs:221]: Could you please explain more? location_to_asset is a Vec and function it iterating over it to find a match. :thinking:

File: zrml/market-commons/src/migrations.rs
Comment (vivekvpandya) [reply zrml/market-commons/src/migrations.rs:162]: took the code from similar old migration, but you are right this if let not relevant in this migration.

File: zrml/market-commons/src/migrations.rs
Comment (vivekvpandya) [reply zrml/market-commons/src/migrations.rs:142]: okay, are there any other places in code too where we should use MarketCommonsPalletApi to make sure we are reading correct storage item? like `on_runtime_upgrade()` or `pre_upgrade()` ?

File: zrml/market-commons/src/migrations.rs
Comment (vivekvpandya) [reply zrml/market-commons/src/migrations.rs:152]: is it fine if I add market_id in panic message like below?
```rust
                   Self::get_temp_storage(&format!("{:?}", market_id.as_slice()))                                                       
                     .expect(&format!("legacy market not found for market_id {:?}", market_id)); 
```

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (maltekliemann) [reply zrml/prediction-markets/src/orml_asset_registry.rs:221]: Yes, but `location_to_asset` is initialized as `Vec::new()` and it never changes, right?

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (maltekliemann) [reply zrml/prediction-markets/src/orml_asset_registry.rs:28]: Sorry, correction: The copyright notice has to stay, of course. Only the license statement can be removed.

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (vivekvpandya) [reply zrml/prediction-markets/src/orml_asset_registry.rs:221]: yes that is correct, but may be it will change in future PR when we add more tests. So the above code is correct right?

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (vivekvpandya) [reply zrml/prediction-markets/src/orml_asset_registry.rs:28]: Yes , I think you already covered that https://github.com/zeitgeistpm/zeitgeist/pull/894#discussion_r1059453184

File: zrml/market-commons/src/migrations.rs
Comment (maltekliemann) [reply zrml/market-commons/src/migrations.rs:142]: Well, in `pre_upgrade` you're writing `LegacyMarket` objects, you don't have any other choice there. In the actualy migration, I think it's fine to use `put_storage_value` even when writing `Market` objects, because we're actually testing that code. My only concern is that tests shouldn't be "self-referential" (not sure if that's even the right word here). I believe one of my recent migrations had that problem.

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (maltekliemann) [reply zrml/prediction-markets/src/orml_asset_registry.rs:221]: As far as I can tell, this is perfectly fine as a mock object. I was just wondering what the point of having `location_to_asset` was. It makes sense to keep it around though.

Did you ever figure out what the point of the `__private` prefix was?

**Review by maltekliemann — CHANGES_REQUESTED** [review]
Thanks for merging the two migrations. Now that I see the diff, it's definitely the better approach. After an in-depth review, some additional problems have popped up:

- The base asset is used as collateral, so it must be used when buying/selling complete sets and when redeeming tokens. I've made a PR which (hopefully) covers all of this: https://github.com/zeitgeistpm/zeitgeist/compare/issue-226...issue-226-mkl-suggestion?diff=split. Please double check that it does before applying these changes.
- This oversight is not covered by our tests. That's very concerning. We need tests for the following scenarios:
  - Editing a market fails due to incorrect base asset (not registered/not allowed).
  - All tests that involve buying/selling complete sets and redeeming. You can probably use `test_case` to save _a lot_ of work and code duplications.
  - Destroying a market slashes the market account correctly.
  
  Let me know if you need any help with that.
- There are two open issues (see above) that should be resolved before we merge.

**Comment by vivekvpandya** [comment]
> The base asset is used as collateral, so it must be used when buying/selling complete sets and when redeeming tokens. I've made a PR which (hopefully) covers all of this: https://github.com/zeitgeistpm/zeitgeist/compare/issue-226...issue-226-mkl-suggestion?diff=split. Please double check that it does before applying these changes.

I agree to this issue, I was thinking on same line so indeed we need a test for illustrating this fact.

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (vivekvpandya) [reply zrml/prediction-markets/src/orml_asset_registry.rs:210]: For `__private` it just seems some coding convention see https://stackoverflow.com/questions/73544273/having-both-public-and-private-code-shared-between-multiple-binaries-rust

**Review by Chralt98 — COMMENTED** [review]
Apart from the comments of you, Malte, I could not find anything. @sea212 ready for your final review. But Harald, maybe you want to wait until the suggestions of Malte are applied.

**Review by maltekliemann — CHANGES_REQUESTED** [review]
I'm still missing some tests here. Everything that has changed must be tested. So what did we change?

- `buy_complete_set`/`sell_complete_set` need to be tested. It's probably easiest to take the tests that already exist and use `test_case` to abstract the base asset.
- The asset used to collateralize complete sets is now not necessarily the same as the asset used for bonds. It would therefore be prudent to verify that whenever bonds

File: zrml/prediction-markets/Cargo.toml
Comment (sea212) [inline zrml/prediction-markets/Cargo.toml:73]:
```suggestion
    "orml-asset-registry?/std",
```

File: zrml/market-commons/src/migrations.rs
Comment (sea212) [inline zrml/market-commons/src/migrations.rs:74]: What do we need this for? I can't find where this is used.

```suggestion
```

File: zrml/prediction-markets/src/lib.rs
Comment (sea212) [inline zrml/prediction-markets/src/lib.rs:56]:
```suggestion
    #[cfg(feature = "parachain")]
    use {
        orml_traits::asset_registry::Inspect,
        zeitgeist_primitives::types::CustomMetadata,
    };
```

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (sea212) [inline zrml/prediction-markets/src/orml_asset_registry.rs:252]: This is actually not necessary, we can achieve everything necessary with the asset registry pallet.
However, we still use the version `polkadot-v0.9.26` on asset registry, which unfortunately does not allow initializing genesis storage (we'd have to write raw keys and values...). So for now, this obviously is a won't fix. However, [from `polkadot-v0.9.29` onwards](https://github.com/open-web3-stack/open-runtime-module-library/compare/polkadot-v0.9.26...polkadot-v0.9.29#diff-973d213acaf59102ed6f6329665a78d77621db8e48f20758a6fdb9be6c75e675R114-R130), we can use the asset registry pallet for that. Can you create an issue to track that please?

File: zrml/market-commons/src/migrations.rs
Comment (vivekvpandya) [reply zrml/market-commons/src/migrations.rs:74]: If we keep it around it may be just hand for next time when we write migration. However I can remove it if you prefer that.

Comment (Chralt98) [reply zrml/market-commons/src/migrations.rs:74]: I think it's okay to leave it in, because of the common use. That's why we used a lot of `#[allow(unused)]`.

File: zrml/prediction-markets/src/orml_asset_registry.rs
Comment (Chralt98) [reply zrml/prediction-markets/src/orml_asset_registry.rs]: https://github.com/zeitgeistpm/zeitgeist/issues/936

File: zrml/prediction-markets/src/tests.rs
Comment (Chralt98) [reply zrml/prediction-markets/src/tests.rs:3476]: https://github.com/zeitgeistpm/zeitgeist/pull/894/commits/5fb067587c8e407b4508e7f9347f820e54cd72b8

File: zrml/prediction-markets/src/tests.rs
Comment (Chralt98) [reply zrml/prediction-markets/src/tests.rs:876]: https://github.com/zeitgeistpm/zeitgeist/pull/894/commits/5fb067587c8e407b4508e7f9347f820e54cd72b8

**Review by Chralt98 — DISMISSED** [review]

**Review by maltekliemann — DISMISSED** [review]
Checks out from my side. Let's wait for @sea212's review.

@saboonikhil Do you need any additional info in the events to track balances of foreign assets that are traded on our chain?

**Comment by saboonikhil** [comment]
> Checks out from my side. Let's wait for @sea212's review.
> 
> @saboonikhil Do you need any additional info in the events to track balances of foreign assets that are traded on our chain?

Just went through the changes.

Market structure already has `base_asset` parameter which now take its effect.

As of now, balances pallet records change in the Ztg balance for an account and tokens pallet for outcome assets, pool shares etc. Will changes to `base_asset`  also come under balances pallet?

**Comment by vivekvpandya** [comment]
> > Checks out from my side. Let's wait for @sea212's review.
> > @saboonikhil Do you need any additional info in the events to track balances of foreign assets that are traded on our chain?
> 
> Just went through the changes.
> 
> Market structure already has `base_asset` parameter which now take its effect.
> 
> As of now, balances pallet records change in the Ztg balance for an account and tokens pallet for outcome assets, pool shares etc. Will changes to `base_asset` also come under balances pallet?

if base_asset is Ztg then it is pallet_balances, other wise it is orm_tokens.

**Comment by saboonikhil** [comment]
> > > Checks out from my side. Let's wait for @sea212's review.
> > > @saboonikhil Do you need any additional info in the events to track balances of foreign assets that are traded on our chain?
> > 
> > 
> > Just went through the changes.
> > Market structure already has `base_asset` parameter which now take its effect.
> > As of now, balances pallet records change in the Ztg balance for an account and tokens pallet for outcome assets, pool shares etc. Will changes to `base_asset` also come under balances pallet?
> 
> if base_asset is Ztg then it is pallet_balances, other wise it is orm_tokens.

Alright, looking forward to see them on testnet 😎

File: zrml/prediction-markets/src/tests.rs
Comment (Chralt98) [inline zrml/prediction-markets/src/tests.rs]: Could you kindly explain why you removed this test?

File: zrml/prediction-markets/src/tests.rs
Comment (Chralt98) [inline zrml/prediction-markets/src/tests.rs]: Could you kindly explain why you removed that test?

File: zrml/prediction-markets/src/tests.rs
Comment (vivekvpandya) [reply zrml/prediction-markets/src/tests.rs]: main branch does not have these tests. https://github.com/zeitgeistpm/zeitgeist/commit/e991d9edbc76db6f2ecc2780a7caf4a4702cff6a#diff-ad4c4ea3d698c97bb435c01c3c30e79c5a60b34af1448fe64b7c87017054b4c6L2248

File: zrml/prediction-markets/src/tests.rs
Comment (Chralt98) [inline zrml/prediction-markets/src/tests.rs:378]: It could be useful here to have a method, which actually fills in the data to the mock asset registry, so that these comments are not required anymore. Just say: `I think it's fine`, if you don't like to change it here. I am okay with this too.

File: zrml/prediction-markets/src/tests.rs
Comment (Chralt98) [inline zrml/prediction-markets/src/tests.rs:3427]: Ok, edit market check is here.

**Review by Chralt98 — COMMENTED** [review]
Looks pretty good to allow different assets as the denominator in prediction markets.

File: zrml/prediction-markets/src/tests.rs
Comment (Chralt98) [reply zrml/prediction-markets/src/tests.rs]: I see, I even think that I deleted them.. But it had the reason that those tests were duplicated. Thanks for the investigation.

**Comment by Chralt98** [comment]
Really great work here Vivek! Love that.

File: zrml/market-commons/src/migrations.rs
Comment (sea212) [reply zrml/market-commons/src/migrations.rs:74]: We have exactly the same code in the prediction markets pallet. We should deduplicate it.

**Review by sea212 — DISMISSED** [review]

**Review by maltekliemann — CHANGES_REQUESTED** [review]
- It seems like the CI doesn't run the tests with `parachain`; it only runs `check`. I guess you could add
  ```
  cargo test -p zrml-prediction-markets --features parachain
  ```

  to `misc.sh` (not sure what the exact structure of those scripts is).
- One test is still failing.
- Is there a test for `create_market` failing if an invalid base asset is specified?
- There are some comments above that are still unaccounted for.

**Comment by vivekvpandya** [comment]
> * Is there a test for `create_market` failing if an invalid base asset is specified?

please check test create_market_with_foreign_assets() it covers to failure where if a FA is not registered in AssetRegistry and if FA is registered but in metadata its not allowed to be used as base_asset.

File: zrml/prediction-markets/src/tests.rs
Comment (vivekvpandya) [reply zrml/prediction-markets/src/tests.rs:378]: Its added in GenesisConfig, similar to how we add default balances to account, otherwise we need to mint balances in very test before we do anything. So I think this fine as comment will guide reader to mock registry.

File: zrml/prediction-markets/src/tests.rs
Comment (Chralt98) [inline zrml/prediction-markets/src/tests.rs:3880]: I see, the Bonds are always taken from ZTG, but not in the base asset.

**Review by Chralt98 — APPROVED** [review]
LGTM

File: scripts/tests/misc.sh
Comment (Chralt98) [reply scripts/tests/misc.sh:14]: No then it's fine!

**Review by maltekliemann — COMMENTED** [review]
Looks good now.

Just one question (I may have asked this before): Why are we using the `parachain` feature to begin with? **Edit.** Seems to me like `parachain` can be removed from `zrml_prediction_markets`. Only price to pay is longer compiler times since you need some XCM imports. Seems worth it, though. Am I wrong about this?

**Comment by vivekvpandya** [comment]
> Looks good now.
> 
> Just one question (I may have asked this before): Why are we using the `parachain` feature to begin with? **Edit.** Seems to me like `parachain` can be removed from `zrml_prediction_markets`. Only price to pay is longer compiler times since you need some XCM imports. Seems worth it, though. Am I wrong about this?

Perhaps it would also bring in some testing time too as there will be XCM tests.
But I think that can be done in separate PR. I think it is better to conclude this PR at this point.

**Review by maltekliemann — APPROVED** [review]

**Pull request merged**
